dm/syncer/error.go

```diff
@@ -211,13 +216,53 @@ func (s *Syncer) handleSpecialDDLError(tctx *tcontext.Context, err error, ddls [
 	}
 	// TODO: add support for downstream alter pk without schema
 
+	// it handles the operations for DDL when encountering `invalid connection` by waiting the asynchronous ddl to synchronize
+	waitAsyncDDL := func(tctx *tcontext.Context, err error, ddls []string, index int, conn *dbconn.DBConn, createTime int64) error {
+		if len(ddls) == 0 || index > len(ddls)-1 || errors.Cause(err) != mysql.ErrInvalidConn || createTime == -1 {
+			return err // return the original error
+		}
+
+		duration := 30
+		failpoint.Inject("ChangeDuration", func() {
+			duration = 1
+		})
+		ticker := time.NewTicker(time.Duration(duration) * time.Second)
+		defer ticker.Stop()
+
+		for {
+			status, err2 := getDDLStatusFromTiDB(tctx, conn, ddls[index], createTime)
+			if err2 != nil {
+				s.tctx.L().Warn("error when getting DDL status fromTiDB", zap.String("error", err2.Error()))
+			}
+			failpoint.Inject("TestStatus", func(val failpoint.Value) {
+				status = val.(string)
+				s.tctx.L().Info("injected test status:", zap.String("TestStatus", status))
+			})
+			switch status {
+			case model.JobStateDone.String(), model.JobStateSynced.String():
```

Comment:
Can we return JobState in getDDLStatusFromTiDB rather then string?

Reply:
I think it may not be necessary. In `getDDLStatusFromTiDB` the original retrieved status is in the form of string, we do not need to manually convert it to `JobState`, which may slow down the performance. Besides, if it cannot find the status in downstream, currently it would return "", but if we change the return type to `JobState`, we may not be able to find a suitable `JobState` to return.